src/index.js

```diff
@@ -11,45 +45,27 @@ import { updateSystem, getDeltaMatrix, getCloseIndex } from './utils';
  * @param {number} [options.tolerance = 0.001] - Factor of the sum of absolute value of original data, to compute stop criterion
  * @param {Array<number>} [options.weights = [1,1,...]] - Initial weights vector, default each point has the same weight
  * @param {number} [options.lambda = 100] - Factor of weights matrix in -> [I + lambda D'D]z = x
- * @param {Array<number>} [options.controlPoints = []] - Array of x axis values to force that baseline cross those points.
- * @param {Array<number>} [options.baseLineZones = []] - Array of x axis values (as from - to), to force that baseline cross those zones.
+ * @param {Array<number>} [options.controlPoints = []] - Array of 0|1 to force the baseline cross those points.
+ * @param {Array<number>} [options.zones = []] - Array of x axis values (as from - to), to force that baseline cross those zones.
  * @returns {{corrected: Array<number>, error: number, iteration: number, baseline: Array<number>}}
  */
 export default function airPLS(x, y, options = {}) {
-  let {
-    maxIterations = 100,
-    lambda = 100,
-    tolerance = 0.001,
-    weights = new Array(y.length).fill(1),
-    controlPoints = [],
-    baseLineZones = [],
-  } = options;
-
-  if (controlPoints.length > 0) {
-    controlPoints.forEach((e, i, arr) => (arr[i] = getCloseIndex(x, e)));
-  }
-  if (baseLineZones.length > 0) {
-    baseLineZones.forEach((range) => {
-      let indexFrom = getCloseIndex(x, range.from);
-      let indexTo = getCloseIndex(x, range.to);
-      if (indexFrom > indexTo) [indexFrom, indexTo] = [indexTo, indexFrom];
-      for (let i = indexFrom; i < indexTo; i++) {
-        controlPoints.push(i);
-      }
-    });
-  }
+  const { weights, controlPoints } = getControlPoints(x, y, options);
+  let { maxIterations = 100, lambda = 10, tolerance = 0.001 } = options;
 
   let baseline, iteration;
-  let nbPoints = y.length;
-  let l = nbPoints - 1;
   let sumNegDifferences = Number.MAX_SAFE_INTEGER;
   let stopCriterion = tolerance * y.reduce((sum, e) => Math.abs(e) + sum, 0);
```

Comment:
This is super slow code. How big is y ?

Reply:
what do you mean with slow?, the size of y depends, usually 128K or more.